File: src/test/groovy/tk/aizydorczyk/kashubian/TransactionSupport.java
package tk.aizydorczyk.kashubian;

import org.springframework.boot.test.context.TestComponent;
import org.springframework.transaction.annotation.Transactional;

import java.util.function.Supplier;

@TestComponent
public class TransactionSupport {
    @Transactional
    public <ResultType> ResultType executeInTransaction(Supplier<ResultType> supplier) {
        return supplier.get();
    }
}

File: src/main/kotlin/tk/aizydorczyk/kashubian/infrastructure/TransactionSupport.kt
package tk.aizydorczyk.kashubian.infrastructure

import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Propagation
import org.springframework.transaction.annotation.Transactional

@Component
class TransactionSupport {
    @Transactional(propagation = Propagation.REQUIRES_NEW)
    fun <ResultType> executeInTransaction(function: () -> ResultType): ResultType {
        return function.invoke()
    }

    @Transactional(readOnly = true)
    fun <ResultType> executeInReadOnlyTransaction(function: () -> ResultType): ResultType {
        return function.invoke()
    }
}